src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/StackTraceMethodMappingNode.cs

```diff
@@ -57,10 +59,86 @@ public override ObjectData GetData(NodeFactory factory, bool relocsOnly = false)
             RelocType reloc = factory.Target.Abi == TargetAbi.CoreRT ?
                 RelocType.IMAGE_REL_BASED_RELPTR32 : RelocType.IMAGE_REL_BASED_ADDR32NB;
 
-            foreach (var mappingEntry in factory.MetadataManager.GetStackTraceMapping(factory))
+            IEnumerable<MetadataMapping<MethodDesc>> mappingEntries = factory.MetadataManager.GetStackTraceMapping(factory);
+            if (!(mappingEntries is ICollection<MetadataMapping<MethodDesc>> collection))
+            {
+                mappingEntries = collection = mappingEntries.ToList();
+            }
+            objData.EmitInt(collection.Count);
+            var reservedSequencePointsOffset = objData.ReserveInt();
+            var reservedStringOffset = objData.ReserveInt();
+
+            int byteCounter = 0;
+            var fileNameToOffset = new Dictionary<string, int>();
+            int GetOrAddStringToHeap(string fileName)
+            {
+                if (!fileNameToOffset.TryGetValue(fileName, out int offset))
+                {
+                    fileNameToOffset.Add(fileName, offset = byteCounter);
+                    byteCounter += sizeof(int) + Encoding.UTF8.GetByteCount(fileName);
+                }
+                return offset;
+            }
+
+            ObjectDataBuilder sequencePointsBuilder = new ObjectDataBuilder(factory, relocsOnly);
+            foreach (var mappingEntry in mappingEntries)
```

Comment:
@jkotas do you have an opinion about potentially using the NativeLayout data structures for this? This appears different enough that I don't see a straightforward mapping, but NativeLayout does have the compressed integer capabilities that we're after.

Reply:
I have been looking around a bit and ofc [927](https://xkcd.com/927/) holds true, so i've found ObjectDataBuilder has [this method](https://github.com/dotnet/corert/blob/1386128b1093e7540daacb1e58fdfcb8f7000a43/src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/ObjectDataBuilder.cs#L138-L167) which only appears to be used for `EHInfo`, System.Reflection.Metadata uses [those](https://github.com/dotnet/corefx/blob/2f9cb4b9caf33a2a6c69d8229c23ed4550079a14/src/System.Reflection.Metadata/src/System/Reflection/Metadata/BlobWriterImpl.cs#L57-L112) for the PDB writing (only supporting signed values up to ~268 million) and BinaryWriter uses this [simple one](https://source.dot.net/#System.Private.CoreLib/shared/System/IO/BinaryWriter.cs,2daa1d14ff1877bd) which would take up 5 bytes for negative numbers but one could shift the sign bit from msb to lsb to combat that.

Reply:
Yes, I think it would make sense to use https://github.com/dotnet/corert/tree/master/src/Common/src/Internal/NativeFormat for this.